lib/path/default.nix

```diff
@@ -149,6 +168,72 @@ in /* No rec! Add dependencies on this file at the top. */ {
           ${subpathInvalidReason subpath}'';
     path + ("/" + subpath);
 
+  /*
+  Remove the first path as a component-wise prefix from the second path.
+  The result is an attribute set containing these attributes:
+
+  - `success` (boolean): Whether the first path is a component-wise prefix of the second path.
+    This is the same as `lib.path.hasPrefix`.
+    If this is `false`, the other attributes will throw an error when evaluated.
+
+  - `components` (list of strings): The resulting path components after removing the prefix.
+    If the first path is the same as the second path, this value is `[ ]`.
+
+  - `subpath` (string): The result as a normalised subpath string, see `lib.path.subpath.normalise`
+    This is the same as `lib.path.subpath.join components`.
```

Comment:
This is clever but cumbersome, and not more convenient than 

```nix
if hasPrefix a b
  then subpath.join (removePrefix a b)
  else throw "oopsie"
```

We can still keep the good error messages in case people don't bother to do the check on their own.

If you're worried about consistency with passing around components instead of a subpath string – well, there is the type signature and one can always add a doc comment stating this is more efficient and one can easily bake it back together into a string with `subpath.join`.

An alternative is providing a "convenience" layer such as `removePrefixAsSubpath` (giving the one doing more work a longer name), but this is more to type than the explicit join.

Reply:
>     lib.path.parts :: Path -> {
>                     root :: FilesystemRoot;
>                     components :: [ Component ];
>                     subpath :: Subpath;
>                   }

This is what's currently called `deconstructPath` somewhere internally, right?

There could be two sets of APIs, one for component-wise paths and one for strings. We also originally discussed a structured representation that is string-coercible, but you deemed that too expensive performance-wise.

I'm personally inclined to keep the interface *and* implementation simple even at the potential cost of performance at scale. We have nice primitives that would make additions almost trivial, and I think we should use them. If that indeed becomes a real performance problem, one can still go measure and figure out alternatives. For now I'd prefer handling strings as we started out with. It's also more predictable for the user, as at each step you'd get out what you expect a path to look like. Otherwise we're opening up the can of design worms all over again.

Reply:
> The problem is that it's not clear what `removePrefix` should return.

I think I'd expect a subpath because these functions have string counterparts.

I agree with @fricklerhandwerk that it's ok to have more functions.

> potential cost of performance

I was going to say that `__toString` is cheaper than `outPath`, but I guess we're dealing with a list :clown_face: 
Just an `Attr` of 24 bytes added to an existing contiguous array and no extra thunk if `__toString` doesn't have the value in its scope, but rather uses the argument. iirc.

Reply:
Fair enough, I now changed this function to return a subpath string. I also created https://github.com/NixOS/nixpkgs/pull/242695 to get the components of a subpath.